internal/databases/mongo/binary/backup.go

```diff
@@ -0,0 +1,319 @@
+package binary
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/wal-g/tracelog"
+	"github.com/wal-g/wal-g/internal"
+	"github.com/wal-g/wal-g/internal/databases/mongo/models"
+	"github.com/wal-g/wal-g/utility"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+const backupType = "binary"
+
+type BackupService struct {
+	Context       context.Context
+	MongodService *MongodService
+	LocalStorage  *LocalStorage
+	BackupStorage *BackupStorage
+
+	MongodBackupMeta *MongodBackupMeta
+
+	BackupDirectoriesMap map[string]*BackupDirectoryMeta
+	BackupFilesMap       map[string]*BackupFileMeta
+}
+
+func GenerateNewBackupName() string {
+	return backupType + "_" + utility.TimeNowCrossPlatformUTC().Format(utility.BackupTimeFormat)
+}
+
+func CreateBackupService(ctx context.Context, mongodService *MongodService, localStorage *LocalStorage,
+	backupStorage *BackupStorage) (*BackupService, error) {
+
+	return &BackupService{
+		Context:       ctx,
+		MongodService: mongodService,
+		LocalStorage:  localStorage,
+		BackupStorage: backupStorage,
+
+		BackupDirectoriesMap: map[string]*BackupDirectoryMeta{},
+		BackupFilesMap:       map[string]*BackupFileMeta{},
+	}, nil
+}
+
+func (backupService *BackupService) DoBackup(backupName string, permanent bool) (err error) {
+	err = backupService.InitializeMongodBackupMeta(backupName, permanent)
+	if err != nil {
+		return err
+	}
+
+	backupCursor, err := backupService.MongodService.GetBackupCursor()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			closeErr := backupCursor.Close(backupService.Context)
+			tracelog.ErrorLogger.Printf("Unable to close backup cursor: %+v", closeErr)
```

Comment:
You can use `utility.LoggedClose`

Reply:
yes, I know it and use it, but in this case backupCursor have Close method with arg (LoggedClose can call Close method without arg)